fedimint-cli/src/client.rs

```diff
@@ -800,13 +808,46 @@ async fn get_note_summary(client: &ClientHandleArc) -> anyhow::Result<serde_json
     .unwrap())
 }
 
+/// Returns a gateway to be used for a lightning operation. If `force_internal`
+/// is true and no `gateway_id` is specified, no gateway will be selected.
 async fn get_gateway(
     client: &ClientHandleArc,
     gateway_id: Option<secp256k1::PublicKey>,
-) -> Option<LightningGateway> {
+    force_internal: bool,
+) -> anyhow::Result<Option<LightningGateway>> {
     let lightning_module = client.get_first_module::<LightningClientModule>();
-    let gateway_id = gateway_id.or(None)?;
-    lightning_module.select_gateway(&gateway_id).await
+    match gateway_id {
+        Some(gateway_id) => {
+            if let Some(gw) = lightning_module.select_gateway(&gateway_id).await {
+                Ok(Some(gw))
+            } else {
+                // Refresh the gateway cache in case the target gateway was registered since the
+                // last update.
+                lightning_module.update_gateway_cache(false).await?;
+                Ok(lightning_module.select_gateway(&gateway_id).await)
+            }
+        }
+        None if !force_internal => {
+            let mut gateways = lightning_module.list_gateways().await;
+            if gateways.is_empty() {
+                // Refresh the gateway cache to find a random gateway to select from.
+                lightning_module.update_gateway_cache(false).await?;
+                gateways = lightning_module.list_gateways().await;
+            }
```

Comment:
you might have expired gateway, always updating is good enough here

Reply:
Do we care about performance? That will introduce another network request for every payment. I suppose the user can always specify `gateway_id` to avoid that.